ui/src/contexts/MeichainContext.js

```diff
@@ -0,0 +1,235 @@
+
+const MeichainContext = createContext()
+
+export const MeichainProvider = ({ children}) => {
+  const [privateKey, setPrivateKey] = useState('')
+  const [meiAddress, setMeiAddress] = useState('')
+
+  const cosmosjs = require("@cosmostation/cosmosjs")
+  const meichain = cosmosjs.network(getMeichainRestServer(), MEICHAIN_CHAIN_ID)
+
+  const isInitiateMeichain = () => {
+    if (!meichain) {
+      throw Error(`Error meichain js is not initialized`)
+    }
+  }
+
+  const getMeichainAddress = (mnemonic) => {
+    try {
+      isInitiateMeichain()
+      const address = meichain.getAddress(mnemonic)
+      setMeiAddress(address)
+      return address
+    } catch (error) {
+      console.log(error)
+      throw Error(`Error cannot get meichain account from mnemonic: ${error.message}`)
+    }
+  }
+
+  const setPrivateKeyFromMnemonic = (mnemonic) => {
+    try {
+      const privateKey = meichain.getECPairPriv(mnemonic)
+      setPrivateKey(privateKey)
+    } catch (error) {
+      throw Error(`Error cannot set private key from mnemonic: ${error.message}`)
+    }
+  }
+
+  const lockCollateral = (amount) => {
+    isInitiateMeichain()
+    if (!privateKey) {
+      throw `Please connect wallet before lock atom`
+    }
+    meichain.getAccounts(meiAddress).then(data => {
+      let stdSignMsg = meichain.newStdMsg({
+        msgs: [
+          {
+            type: "meichain/LockCollateral",
+            value: {
+              Amount: String(amount),
+              Sender: meiAddress,
+            }
+          }
+        ],
+        chain_id: MEICHAIN_CHAIN_ID,
+        fee: { amount: [], gas: String(200000) },
+        memo: "",
+        account_number: String(data.result.value.account_number),
+        sequence: String(data.result.value.sequence)
+      });
+      let signedTx = meichain.sign(stdSignMsg, privateKey)
+      signedTx = convertSignMsg(signedTx)
+      meichain.broadcast(signedTx).then(response => console.log(response));
+    })
+  }
+
+  const unlockCollateral = (amount) => {
+    isInitiateMeichain()
+    if (!privateKey) {
+      throw `Please connect wallet before unlock atom`
+    }
+    meichain.getAccounts(meiAddress).then(data => {
+      let stdSignMsg = meichain.newStdMsg({
+        msgs: [
+          {
+            type: "meichain/UnlockCollateral",
+            value: {
+              Amount: String(amount),
+              Sender: meiAddress,
+            }
+          }
+        ],
+        chain_id: MEICHAIN_CHAIN_ID,
+        fee: { amount: [], gas: String(200000) },
+        memo: "",
+        account_number: String(data.result.value.account_number),
+        sequence: String(data.result.value.sequence)
+      });
+    
+      let signedTx = meichain.sign(stdSignMsg, privateKey);
+      signedTx = convertSignMsg(signedTx)
+      meichain.broadcast(signedTx).then(response => console.log(response))
+    })
+  }
```

Comment:
The sign and broadcast are duplicated. Make it as helper function.

Reply:
fixed